crates/rpc/rpc-types-compat/src/transaction.rs

```diff
@@ -46,3 +59,207 @@ pub trait TransactionCompat: Send + Sync + Unpin + Clone + Debug {
+            }),
+            tx_info,
+        )
+    }
+}
+
+/// Adds extra context to [`TransactionInfo`].
+pub trait TxInfoMapper<T> {
+    /// An associated output type that carries [`TransactionInfo`] with some extra context.
+    type Out;
+    /// An associated error that can occur during the mapping.
+    type Err;
+
+    /// Performs the conversion.
+    fn try_map(&self, tx: T, tx_info: TransactionInfo) -> Result<Self::Out, Self::Err>;
+}
+
+impl<T> TxInfoMapper<&T> for () {
+    type Out = TransactionInfo;
+    type Err = Infallible;
+
+    fn try_map(&self, _tx: &T, tx_info: TransactionInfo) -> Result<Self::Out, Self::Err> {
+        Ok(tx_info)
+    }
+}
+
+/// Creates [`OpTransactionInfo`] by adding [`OpDepositInfo`] to [`TransactionInfo`] if `tx` is a
+/// deposit.
+pub fn try_into_op_tx_info<T: ReceiptProvider<Receipt: DepositReceipt>>(
+    provider: &T,
+    tx: &OpTxEnvelope,
+    tx_info: TransactionInfo,
+) -> Result<OpTransactionInfo, ProviderError> {
+    let deposit_meta = if tx.is_deposit() {
+        provider.receipt_by_hash(tx.tx_hash())?.and_then(|receipt| {
+            receipt.as_deposit_receipt().map(|receipt| OpDepositInfo {
+                deposit_receipt_version: receipt.deposit_receipt_version,
+                deposit_nonce: receipt.deposit_nonce,
+            })
+        })
+    } else {
+        None
+    }
+    .unwrap_or_default();
+
+    Ok(OpTransactionInfo::new(tx_info, deposit_meta))
+}
+
+impl IntoRpcTx<op_alloy_rpc_types::Transaction> for OpTxEnvelope {
+    type TxInfo = OpTransactionInfo;
+
+    fn into_rpc_tx(
+        self,
+        signer: Address,
+        tx_info: OpTransactionInfo,
+    ) -> op_alloy_rpc_types::Transaction {
+        op_alloy_rpc_types::Transaction::from_transaction(self.with_signer(signer), tx_info)
+    }
+}
+
+impl TryIntoSimTx<EthereumTxEnvelope<TxEip4844>> for TransactionRequest {
+    fn try_into_sim_tx(self) -> Result<EthereumTxEnvelope<TxEip4844>, ValueError<Self>> {
+        Self::build_typed_simulate_transaction(self)
+    }
+}
+
+impl TryIntoSimTx<OpTxEnvelope> for TransactionRequest {
+    fn try_into_sim_tx(self) -> Result<OpTxEnvelope, ValueError<Self>> {
+        let request: OpTransactionRequest = self.into();
+        let tx = request.build_typed_tx().map_err(|request| {
+            ValueError::new(request.as_ref().clone(), "Required fields missing")
+        })?;
+
+        // Create an empty signature for the transaction.
+        let signature = Signature::new(Default::default(), Default::default(), false);
+
+        Ok(tx.into_signed(signature).into())
+    }
+}
+
+/// Conversion into transaction RPC response failed.
+#[derive(Debug, Clone, Error)]
+#[error("Failed to convert transaction into RPC response")]
+pub struct TransactionConversionError;
+
+/// Generic RPC response object converter for primitives `N` and network `E`.
+#[derive(Debug)]
+pub struct RpcTransactionConverter<N, E, Err, Map = ()> {
```

Comment:
should we perhaps just name this `RpcConverter`, we kinda only need to handle transaction types so perhaps `RpcTransactionConverter` is already everything we need

Reply:
Nice!